tests/shared/process-event.ts

```diff
@@ -47,7 +48,31 @@ async function createPerson(
     return server.db.createPerson(DateTime.utc(), properties, team.id, null, false, new UUIDT().toString(), distinctIds)
 }
 
-type ReturnWithHub = { hub?: Hub; closeHub?: () => Promise<void> }
+export type ReturnWithHub = { hub?: Hub; closeHub?: () => Promise<void> }
+
+export const getEventsByPerson = async (hub: Hub) => {
+    // Helper function to retrieve events paired with their associated distinct
+    // ids
+    const persons = await hub.db.fetchPersons()
+    const events = await hub.db.fetchEvents()
+
+    return await Promise.all(
+        persons.map(async (person) => {
+            const distinctIds = await hub.db.fetchDistinctIdValues(person)
+
+            return [
+                distinctIds,
+                (events as Event[])
+                    .filter((event) => distinctIds.includes(event.distinct_id))
+                    .map((event) => event.event),
+            ] as const
+        })
+    )
+}
+
+export const createIdentifiedPerson = async (hub: Hub, teamId: number, distinctId: string): Promise<void> => {
```

Comment:
Is there any reason to reach into the internals here? It would seem that we can get to the same state just with a sequence of events and avoid artificially updating the db.

Reply:
The description is coming soon and I was going to comment on this specific line.  

The tests you wrote were an amazing foundation here, so big thanks for that.

However, there was a conceptual mistake with the $identify tests.

The following:

```js
posthog.identify('first_id')
posthog.capture('some event')
posthog.identify('second_id')
posthog.capture('some other event')
```

Leads to and should lead to **one** person in the DB. Reason being that we've never seen `second_id` before, so we actually just pop that ID into the person from `first_id`, we never actually merge _persons_.

So to actually test that we're not allowing the merging of persons here, we need a person to exist, and it's just cleaner/easier to tap into the internals for it.

I completely get the sentiment of mimicking the libs here, but this does in fact do that. It works with the (very reasonable) assumption that by the time this happens:

```js
posthog.identify('first_id')
posthog.capture('some event')
posthog.identify('second_id')
posthog.capture('some other event')
```

somewhere else `second_id` has already been created. The assumption is actually more reasonable than doing it all without internals because it's much more likely that someone ends up trying to merge into a person that already exists than they create 2 new persons in a session (this would require some odd circumstances).

Hope this makes sense?

Reply:
If it's not a valid sequence of events, then I would suggest we get rid and replace with one that is valid.

Point being, it's very easy to test how the system behaves in response to events, and these are going to be pretty stable and representative of actual interactions (as this is what we tell others to use), where as reaching in we are making it harder to refactor.

Reply:
Right, but the thing is - some scenarios are more likely given the current state of our data than through a sequence in a session.

I'm happy to make some changes here but ultimately I see it as valid to have the DB in a given state and trigger events from there to see how we react given the current state.

And what we generally see is:

User A opens up their app and a person is created for them.

User B opens up their app and a person is created for them, which is then asked to be merged into User A.

These are not a sequence of events happening at once.

Reply:
But yes indeed they are a sequence of events - they're just not very compatible with the stateful mock client that's there, but I can certainly re-write/circumvent that too

Reply:
I think where ever we can do a valid sequence of events in and validate that the state of the system is correct that should be the preference. For the case you're mentioned, that seems like you've basically got a valid sequence of events there.

I guess there are cases where we can't get the db into a specific state through events so I guess in these cases we don't have much choice.

Reply:
updated to reflect this - thanks for pushing me here!